source/adapters/level_zero/context.hpp

```diff
@@ -333,10 +334,38 @@ struct ur_context_handle_t_ : _ur_object {
         return WithProfiling ? &EventCaches[2] : &EventCaches[3];
       }
     }
+  };
+  auto getCounterBasedEventCache(bool UsingImmediateCmdList,
+                                 ur_device_handle_t Device) {
+    if (UsingImmediateCmdList) {
+      if (Device) {
+        auto EventCachesMap = &EventCachesDeviceMap[4];
+        if (EventCachesMap->find(Device) == EventCachesMap->end()) {
+          EventCaches.emplace_back();
+          EventCachesMap->insert(
```

Comment:
why not do the insert in emplace back above?
why split it into 2 steps?

Reply:
Emplaces constructs and allocates, insert simply copies. [**[UR] Phase 2 of counter-based event implementation
    
    PR in UR: https://github.com/oneapi-src/unified-runtime/pull/1698**](https://cplusplus.com/forum/general/122033/#google_vignette)

Reply:
emplace_back is for eventcache which only stores the handles to ur_events, where the actual event cache is stored.
insert is for eventCachesDevicesMap, which is for keeping track of which events are associated with devices.